src/ImageSharp/Processing/Processors/Effects/OilPaintingProcessor{TPixel}.cs

```diff
@@ -140,29 +142,29 @@ public void Invoke(in RowInterval rows)
                             int offsetX = x + fxr;
                             offsetX = Numerics.Clamp(offsetX, 0, maxX);
 
-                            Vector4 vector = sourceOffsetRow[offsetX].ToVector4();
+                            Vector4 vector = Vector4.Clamp(sourceOffsetRow[offsetX].ToScaledVector4(), Vector4.Zero, Vector4.One);
```

Comment:
Unfortunately, `ToScaledVector4` doesn't guarantee that the vector will be clamped to `0..1` since a pixel implementation is free to return anything. (In fact `RgbaVector` doesn't clamp!)

With clamping I was able to undo the buffer overallocation in [99771d6](https://github.com/SixLabors/ImageSharp/pull/2519/commits/99771d61a87de9448698aa1368979d5d23e983c6).

I think the PR should be good now if you are also fine with the changes @JimBobSquarePants .

Reply:
Shouldn’t we fix RgbaVector? ScaledVector is supposed to guarantee 0…1

Reply:
Just had a look at several of our pixel formats. None of the sanitize the range in construction. We should fix that across the board.

Reply:
> None of the sanitize the range in construction

I wonder which are the pixel types where `ToScaledVector4()` could go out of range? To me it looks like it can happen only with floating point ones (`RgbaVector`, `HalfVector2|4`, `HalfSingle` ...) Do you see any other ones?

Sanitizing in the constructor is insufficient since (1) there are public fields/properties to manipulate the values (2) low-level `MemoryMarshal.Cast`-like manipulation is completely normal when working with pixel buffers and it can also push values out of range. If we want a guarantee that `ToScaledVector4()` implementations always return values in `0..1`, clipping should happen in `ToScaledVector4()`.

However, even if guarantee this in our own pixel types, arbitrary user pixel types can still potentially violate the contract. If it only leads to an `IndexOutOfRangeException`, we can blame it on the user. However, if we have unsafe indexing depending on the the values returned from `ToScaledVector4()` (like we used to do in OilPaint) we have a serious bug. We *may* have a similar problem here:

https://github.com/SixLabors/ImageSharp/blob/e81efa3ca649116282e3f58e76abeecbd625e482/src/ImageSharp/Processing/Processors/Normalization/AutoLevelProcessor%7BTPixel%7D.cs#L151-L153

My recommendation would be to revisit these processors before doing anything else.

Reply:
All of the pixel formats accepting a `float` lack sanitation so `Short2`, `Short4`, and the `NormalizedXX` variants would be on the list also.

Scaled vector shouldn't clip. It should normalize the scale of the min max values of the pixel format. The mutability of the pixel formats makes sanitation difficult but as you say, avoiding unnecessary unsafe code turns a trivial exception to a serious one so we should definitely review and fix any processors that could be affected.

Regarding this PR. That clamp is going to hurt performance and _should_ be unnecessary given the correct `Vector4` range. I'd say we revert it and merge before continuing to sanitize the processors.

Reply:
See [ac0c1f9](https://github.com/SixLabors/ImageSharp/pull/2519/commits/ac0c1f903d7aa8583bf2eb4e29975f0a88bd5989). With aéé  the confusion this PR started with, it would make sense to squash on merge.